server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,648 @@
+                    sortedSet.lookupOrd(sortedSet.getValueCount() - 1);
+                    break;
+                default:
+                    assert false : "Unknown docValues type [" + dvType + "]";
+                    throw new IllegalStateException("Unknown docValues type [" + dvType + "]");
+            }
+            stats.addDocValues(fieldInfo.name, directory.getBytesRead());
+        }
+    }
+
+    void analyzePostings(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        // TODO: FieldsReader has stats() which might contain the disk usage infos
+        // Also, can we track the byte reader per field extension to avoid visiting terms multiple times?
+        FieldsProducer postingsReader = reader.getPostingsReader();
+        if (postingsReader == null) {
+            return;
+        }
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetBytesRead();
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+
+            // Visit terms index and term dictionary and postings
+            final BytesRefIterator termsIterator = terms.iterator();
+            TermsEnum termsEnum = terms.iterator();
+            BytesRef bytesRef;
+            while ((bytesRef = termsIterator.next()) != null) {
+                cancellationChecker.logEvent();
+                termsEnum.seekExact(bytesRef); // it's expensive to look up every term
+                termsEnum.totalTermFreq();
+                for (long idx = 0; idx <= 8; idx++) {
+                    cancellationChecker.logEvent();
+                    final int skipDocID = Math.toIntExact(idx * reader.maxDoc() / 8);
+                    postings = termsEnum.postings(postings, PostingsEnum.NONE);
+                    if (postings.advance(skipDocID) != DocIdSetIterator.NO_MORE_DOCS) {
+                        postings.freq();
+                        postings.nextDoc();
+                    } else {
+                        break;
+                    }
+                }
+            }
+            final long termsBytes = directory.getBytesRead();
+            stats.addTerms(field.name, termsBytes);
+
+            // Visit positions, offsets, and payloads
+            termsEnum = terms.iterator();
+            while (termsEnum.next() != null) {
+                cancellationChecker.logEvent();
+                termsEnum.docFreq();
+                termsEnum.totalTermFreq();
+                postings = termsEnum.postings(postings, PostingsEnum.ALL);
+                while (postings.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                    if (terms.hasPositions()) {
```

Comment:
should we skip computing disk usage of proximity data entirely when positions are not indexed, ie. not even iterate a second time through the terms dictionary?

Reply:
++, adapted in https://github.com/elastic/elasticsearch/pull/74051/commits/0efab4490ed81f5ad8779e3cdd0010d324098b98.